pinot-query-planner/src/main/java/org/apache/pinot/query/planner/ExplainPlanStageVisitor.java

```diff
@@ -0,0 +1,201 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.planner;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.pinot.core.transport.ServerInstance;
+import org.apache.pinot.query.planner.stage.AggregateNode;
+import org.apache.pinot.query.planner.stage.FilterNode;
+import org.apache.pinot.query.planner.stage.JoinNode;
+import org.apache.pinot.query.planner.stage.MailboxReceiveNode;
+import org.apache.pinot.query.planner.stage.MailboxSendNode;
+import org.apache.pinot.query.planner.stage.ProjectNode;
+import org.apache.pinot.query.planner.stage.SortNode;
+import org.apache.pinot.query.planner.stage.StageNode;
+import org.apache.pinot.query.planner.stage.StageNodeVisitor;
+import org.apache.pinot.query.planner.stage.TableScanNode;
+import org.apache.pinot.query.planner.stage.ValueNode;
+
+
+public class ExplainPlanStageVisitor implements StageNodeVisitor<StringBuilder, ExplainPlanStageVisitor.Context> {
```

Comment:
1. can we implement this not as visitor but as calcite's `RelWriter`?
3. could we add some javadoc?

Reply:
Not sure I understand your suggestion - this visits `StageNode` as opposed to `RelNode`, and it wouldn't be straightforward to make `StageNode` implement `RelNode`. Was that your suggestion?

Can definitely add javadoc.

Reply:
i meant similar to `RelWriter` as it seems like this change adds more lines to the codebases than a simple toString method on QueryPlanner

Reply:
ah I see - I also added some more functionality in this PR, which perhaps is a bit confusing (I added which server a stage was being executed on). I'll see if I can simplify it, but I don't think more LOC is necessarily a bad thing if it's much easier to extend and read.

Reply:
^ that knowledge ( which server a stage was being executed on) was useful i actually checked through that as well!

Reply:
I looked at `RelWriter` and I'm not sure it's complex enough to do what I want (different behaviors based on different `StageNode` types) - the other issue is that some of the information that I want to augment with is not in the node itself (sever info, segment info) so I can't just encapsulate it in the `StageNode#explain` method generically... or I guess I can but I'd need to do more refactoring to get that to work). 

Let's leave that as a follow up to consider - for now I just added the javaodc